src/clustermove.cpp

```diff
@@ -2,61 +2,96 @@
+    auto &json_shape = j["shape distribution"] = json::array();
+    for (const auto &[size, shape_anisotropy] : shape.shape_distribution) {
+        json_shape.push_back({size, shape_anisotropy.avg()});
+    }
+}
+
+FindCluster::FindCluster(Space &spc, const json &j) : spc(spc) {
+    single_layer = j.value("single_layer", false);
+    if (j.count("spread")) {
+        faunus_logger->error("cluster: 'spread' is deprecated, use 'single_layer' instead");
+    }
+    molecule_names = j.at("molecules").get<decltype(molecule_names)>(); // molecule names
+    molids = Faunus::names2ids(Faunus::molecules, molecule_names);      // names --> molids
+    std::sort(molids.begin(), molids.end());
+    updateMoleculeIndex();
+
+    // read satellite ids (molecules NOT to be considered as cluster centers)
+    registerSatellites(j.value("satellites", std::vector<std::string>()));
+    parseThresholds(j.at("threshold"));
+}
+
+double FindCluster::clusterProbability(const FindCluster::Tgroup &group1, const FindCluster::Tgroup &group2) const {
     if (!group1.empty() and !group2.empty()) {
         if (spc.geo.sqdist(group1.cm, group2.cm) <= thresholds_squared(group1.id, group2.id)) {
             return 1.0;
         }
     }
     return 0.0;
 }
-void Cluster::_to_json(json &j) const {
-    using namespace u8;
-    j = {{"dir", translation_direction},
-         {"dp", translation_displacement_factor},
-         {"dprot", rotation_displacement_factor},
-         {"single_layer", single_layer},
-         {"dirrot", rotation_axis},
-         {rootof + bracket("r" + squared), std::sqrt(translational_mean_square_displacement.avg())},
-         {rootof + bracket(theta + squared) + "/" + degrees,
-          std::sqrt(rotational_mean_square_displacement.avg()) / 1.0_deg},
-         {bracket("N"), average_cluster_size.avg()},
-         {"bias rejection rate", double(bias_rejected) / cnt},
-         {"clusterdistribution", cluster_size_distribution}};
-    _roundjson(j, 3);
 
-    // print threshold matrix
-    auto &_j = j["threshold"];
-    for (auto i : molids)
-        for (auto j : molids)
-            if (i >= j) {
-                std::string key = Faunus::molecules[i].name + " " + Faunus::molecules[j].name;
-                _j[key] = std::sqrt(thresholds_squared(i, j));
-                _roundjson(_j[key], 3);
-            }
-
-    // print satellite molecules
-    if (not satellites.empty()) {
-        auto &_j = j["satellites"];
-        _j = json::array();
-        for (auto id : satellites)
-            _j.push_back(Faunus::molecules[id].name);
-    }
+/**
+ * Search for molecules participating in the cluster move.
+ * This should be called for every move event as a
+ * grand canonical move may have changed the number
+ * of particles.
+ *
+ * Atomic groups and inactive groups are ignored.
+ */
+void FindCluster::updateMoleculeIndex() {
+    auto to_group_index = [&](const Tgroup &group) { return &group - &spc.groups.front(); };
```

Comment:
If possible, hiding the pointer arithmetic (black magic) in a named function at groups or similar would be nice. ;)

Reply:
Subtracting two addresses doesn't seem so magical to me ;-)